[Controls] timeslider control

nreese opened the pull request:
PR creates a time slider control that is functionally compatible with existing Maps time slider (with the intent that time slider control will one day replace the Maps time slider).

Key features:
1) Time slider control is not tied to data view or data view field (just like the global time picker)
2) Control updates dashboard container input state with timeslice. Panels can use both timeRange (from global timepicker) and timeslice (from time slider) to display content. Maps embeddable uses both timeRange and timeslice to use client-side masking when all data can be loaded to client. For all other embeddables, timeslice is used instead of timeRange.
3) Control uses observable from dashboard container to play next frame when all panels for the current frame finish loading.

<img width="600" alt="Screen Shot 2022-08-29 at 1 26 38 PM" src="https://user-images.githubusercontent.com/373691/187282240-0dd23a98-c7a8-462f-915c-4c41dacf1389.png">

To test:
1) Create dashboard with lots of different types of embeddables
2) Add timeslider control
3) mess around with control and verify dashboard responds as expected.

nreese commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/kibana-presentation (Team:Presentation)

elasticmachine commented:
Pinging @elastic/kibana-gis (Team:Geo)

nreese commented:
@elasticmachine merge upstream

jsanz commented:
I gave the new control a run against a rather large OSM dataset.

[vokoscreenNG-2022-08-31_19-05-39.webm](https://user-images.githubusercontent.com/188264/187738892-54a2b26e-70fe-44ac-a8b0-449cebc98cf0.webm)

Not being able to select exact time ranges for the steps is a bit frustrating. I wanted to step over years but the dates "drifted" as I could not select an exact upper bound so I could not create frames between January 1st and Dec. 31st.

On the other hand the "jump" in the map legend limits can be confusing but I guess this is totally out of scope for this PR but something we should try to help users somehow address. In fact this issue is present for all the panels, just see the changes over time for the Y axis of the upper histogram.

Finally something that is a bit distracting, sliding the knobs for the upper/lower bounds of the range are updating the dashboard for every mouse move. I think we should add some throttling to the event so the dashboard is not that "jumpy" when interacting with the control.

I find this feature amazing, the interactivity that it provides to our visualizations is so cool.

nreese commented:
> Not being able to select exact time ranges for the steps is a bit frustrating. I wanted to step over years but the dates "drifted" as I could not select an exact upper bound so I could not create frames between January 1st and Dec. 31st.

This is an [existing issue](https://github.com/elastic/kibana/issues/103141) with maps time slider. This PR does not attempt to address existing issues with the maps timeslider. Instead, focusing on migrating the current functionality as-is. Then, we can iteratively enhance the timeslider control to address these issues.

> On the other hand the "jump" in the map legend limits can be confusing but I guess this is totally out of scope for this PR but something we should try to help users somehow address. In fact this issue is present for all the panels, just see the changes over time for the Y axis of the upper histogram.

Maps allows ignoring time range when calculating field statistics so that time range changes will not change scales. You could turn that setting on for your layers. Or we could ignore timeslice for field statistics but I would consider this in a separate issue.

> Finally something that is a bit distracting, sliding the knobs for the upper/lower bounds of the range are updating the dashboard for every mouse move. I think we should add some throttling to the event so the dashboard is not that "jumpy" when interacting with the control.

The control value publishing is debounced and this debounce value could be increased if needed.

kertal reviewed (approved):
DataDiscovery changes LGTM, tested a saved search on a dashboard with time slider control  locally, and it worked like expected 👍 that's a nice new feature ❤️

ThomThomson commented on src/plugins/controls/public/time_slider/components/time_slider.tsx:
```diff
@@ -0,0 +1,89 @@
+    return state.explicitInput.value;
+  });
+  const isOpen = select((state) => {
+    return state.componentState.isOpen;
```
Nice use of `ComponentState` in here - like to see that!

ThomThomson commented on src/plugins/embeddable/public/lib/containers/container.ts:
```diff
@@ -82,6 +93,26 @@ export abstract class Container<
         this.maybeUpdateChildren(currentPanels, prevPanels);
       });
+
+    this.anyChildOutputChange$ = this.getOutput$().pipe(
```
Nice to see this added in a generic place!

nreese commented:
@elasticmachine merge upstream

nickpeihl reviewed (approved):
lgtm! This is a great new control.

My only nits have already been covered by @ThomThomson.

code review and tested in chrome

flash1293 reviewed (approved):
Tested and viseditors changes LGTM, all chart types work well. Great feature!

nreese commented:
@elasticmachine merge upstream

nreese commented:
@ThomThomson 

See below for addressing feedback.

In addition to your feedback, I have updated the time slider to save state as relative to time range so that when restoring a time slider control, the value adjusts for relative time ranges.

**Unsaved Changes**
Resolved with https://github.com/elastic/kibana/pull/139228/commits/54d642550b72a957fcc59cf88de32129b31300ac

**Slider interaction Nit**
As discussed, range slider does not have any special behavior to only apply changes on mouse up. Both range slider and time slider EuiDualRange and debounce publishing value. I am not sure why time slider interaction is not as smooth as range slider interaction.

**Dragging UI Nit**
Resolved with https://github.com/elastic/kibana/pull/139228/commits/29884404fa017db978420c46b6863b5091429e64

**Width Question**
> Do you plan to support any settings on the time slider Control? Specifically a width setting might be helpful, to make it work like the other control types.

No, I think it best to only support "large" time slider width and wait for user requests to support "small" or "medium". The reason being is that the time slider displays ticks which may not display well at "small" and "medium" sizes

**Label position Above Nit**
Resolved with https://github.com/elastic/kibana/pull/139228/commits/c3f5de1987e57657c9b7b2895019c2ced64451ec. Setting title to "Time slider" because something is needed to preserve spacing

**Functional Testing**
Resolved with https://github.com/elastic/kibana/pull/139228/commits/00989a82a8bb501045bc80d3bd1a0f05eb869b90

ThomThomson reviewed (approved):
Re-reviewing, because it looks like all of my major concerns have been addressed. Everything else we can look into as a follow up.

nreese commented:
@teresaalvarezsoler 
> Small bug/edge case that I found, if I add a numeric field control (range slider) after the timeslider and I press the play button while keeping the range slider opened (first image) at some point the range slider disappears (second image).

Thanks for finding this. This is an existing issue with range slider that is easier to trigger with time slider. I have opened an issue for the problem https://github.com/elastic/kibana/issues/140135

andreadelrio commented:
@nreese I have a question, what rule(s) if any is the time slider control following to set its width?

Seems like the behavior now is (see video):

a) if user adds a time slider control as the **first** control in the control group it goes full width
b) user then adds another control type with say "small width - do not expand width to fit available space"
c) user deletes the control from (b)
d) user then deletes the time slider control
e) user re-adds the time slider control and now the width is no longer full width?

https://user-images.githubusercontent.com/4016496/188758691-765454b0-f906-4014-8ff8-d592b1a45898.mov

Or are width settings no longer needed because of the future of Unified Search? (cc @ThomThomson)

nreese commented:
> I have a question, what rule(s) if any is the time slider control following to set its width

Width is just using `large` control CSS rules

nreese commented:
@elasticmachine merge upstream

andreadelrio reviewed (changes requested):
From a UX point of view, I think we should allow users to customize the width of the time slider control but that can likely wait until we implement Controls x Unified Search.

Regarding the current state of this feature, one design change that we would need as a minimum is making sure the behavior of the time slider control's width is consistent. If the time slider control (as it is now with no width settings of its own) is the only item in the control group it should always behave the same. In this case, we can make it so that it always expands to fit available space.

andreadelrio reviewed (approved):
@nreese thanks for making that change, LGTM.
